linalg/slepc.cpp

```diff
@@ -0,0 +1,254 @@
+                         const char help[])
+{
+   ierr = SlepcInitialize(argc,argv,rc_file,help);
+   MFEM_VERIFY(!ierr,"Unable to initialize SLEPc");
+}
+
+void MFEMFinalizeSlepc()
+{
+   ierr = SlepcFinalize();
+   MFEM_VERIFY(!ierr,"Unable to finalize SLEPc");
+}
+
+
+SlepcEigenSolver::SlepcEigenSolver(MPI_Comm comm, const std::string &prefix)
+{
+   clcustom = false;
+   _tol = PETSC_DEFAULT;
+   _max_its = PETSC_DEFAULT;
+   VR = NULL;
+   VC = NULL;
+   operatorset = false;
+
+   ierr = EPSCreate(comm,&eps); CCHKERRQ(comm,ierr);
+   ierr = EPSSetOptionsPrefix(eps, prefix.c_str()); PCHKERRQ(eps, ierr);
+}
+
+SlepcEigenSolver::~SlepcEigenSolver()
+{
+   MPI_Comm comm;
+   ierr = PetscObjectGetComm((PetscObject)eps,&comm); PCHKERRQ(eps,ierr);
+   ierr = EPSDestroy(&eps); CCHKERRQ(comm,ierr);
+}
+
+
+void SlepcEigenSolver::SetOperator(const PetscParMatrix &op)
+{
+   if (operatorset)
+   {
+      delete VR;
+      delete VC;
+      VR = VC = NULL;
+   }
+
+   ierr = EPSSetOperators(eps,op,NULL); PCHKERRQ(eps, ierr);
+   operatorset = true;
+}
+
+void SlepcEigenSolver::SetOperators(const PetscParMatrix &op,
+                                    const PetscParMatrix&opB)
+{
+   if (operatorset)
+   {
+      delete VR;
+      delete VC;
+      VR = VC = NULL;
+   }
+
+   ierr = EPSSetOperators(eps,op,opB); PCHKERRQ(eps,ierr);
+   operatorset = true;
+}
+
+void SlepcEigenSolver::SetTol(double tol)
+{
+   _tol = tol;
+   ierr = EPSSetTolerances(eps,_tol,_max_its); PCHKERRQ(eps,ierr);
+}
+
+void SlepcEigenSolver::SetMaxIter(int max_its)
+{
+   _max_its = max_its;
+   ierr = EPSSetTolerances(eps,_tol,_max_its); PCHKERRQ(eps,ierr);
+}
+
+void SlepcEigenSolver::SetNumModes(int num_eigs)
+{
+   ierr = EPSSetDimensions(eps,num_eigs,PETSC_DECIDE,PETSC_DECIDE);
+   PCHKERRQ(eps,ierr);
+}
+
+void SlepcEigenSolver::Solve()
+{
+   Customize();
+
+   ierr = EPSSolve(eps); PCHKERRQ(eps,ierr);
+}
+
+void SlepcEigenSolver::Customize(bool customize) const
+{
+   if (!customize) {clcustom = true; }
+   if (!clcustom)
+   {
+      ierr = EPSSetFromOptions(eps); PCHKERRQ(eps,ierr);
+   }
+   clcustom = true;
+}
+
+void SlepcEigenSolver::GetEigenvalue(unsigned int i, double & lr) const
+{
+   ierr = EPSGetEigenvalue(eps,i,&lr,NULL); PCHKERRQ(eps,ierr);
+}
+
+void SlepcEigenSolver::GetEigenvalue(unsigned int i, double & lr,
+                                     double & lc) const
+{
+   ierr = EPSGetEigenvalue(eps,i,&lr,&lc); PCHKERRQ(eps,ierr);
+}
+
+void SlepcEigenSolver::GetEigenvector(unsigned int i, Vector & vr) const
+{
+   if (!VR)
+   {
+      Mat pA = NULL;
+      ierr = EPSGetOperators(eps, &pA, NULL); PCHKERRQ(eps,ierr);
+      VR = new PetscParVector(pA, true, false);
+   }
+   VR->PlaceArray(vr.GetData());
+   ierr = EPSGetEigenvector(eps,i,*VR,NULL); PCHKERRQ(eps,ierr);
+   VR->ResetArray();
+
+}
+
+void SlepcEigenSolver::GetEigenvector(unsigned int i, Vector & vr,
+                                      Vector & vc) const
+{
+   if (!VR || !VC)
```

Comment:
Should not needed, since you create the vectors at SetOperators time. Just add
```
MFEM_VERIFY(VR,"Missing right vector")
MFEM_VERIFY(VC,"Missing left vector")
```

Reply:
Done